cpp/doc/TESTING.md

```diff
@@ -0,0 +1,105 @@
+# Unit Testing in libcuspatial
+
+Unit tests in libcuspatial are written using
+[Google Test](https://github.com/google/googletest/blob/master/docs/primer.md).
+
+## Best Practices: What Should We Test?
+
+In general we should test to make sure all code paths are covered. This is not always easy or
+possible. But generally this means we test all supported combinations of algorithms and data types,
+and the main iterator and container types supported by algorithms.  Here are some other guidelines.
+
+ * In general empty input is not an error in libcuspatial. Typically empty input results in empty
+   output. Tests should verify this.
+
+ * Most algorithms should have one or more tests exercising inputs with a large enough number of
+   rows to require launching multiple thread blocks, especially when values are ultimately
+   communicated between blocks (e.g. reductions). This is especially important for custom kernels
+   but also applies to Thrust and CUB algorithm calls with lambdas / functors.
+
+ * Test exceptional cases. For example, anything that causes the function to `throw`.
+
+ * Test boundary cases. For example points that fall exactly on lines or boundaries.
+
```

Comment:
Similar to the above, do we want to specify tests written only for public/non-`detail` APIs, but that those tests should be written such that as close to 100% test coverage of internal functions occurs?

Reply:
This reminds me that `detail` APIs typically include the `mr` and `stream` args which I don't think we are currently testing. Do we want to mention this?

Reply:
Public APIs have MR args if they allocate output memory, so that is tested. All header-only APIs have non-optional stream args. We are moving most testing to the header-only API.

Reply:
Done.